content/dakshdeephere/threading.md

```diff
@@ -0,0 +1,355 @@
+    printf("second thread ret2 = %d \n", ret2);
+
+    return 0;
+}
+
+void *myfunc(void *myvar)
+{
+    char *msg;
+    msg = (char *)myvar;
+    int i;
+
+    for (i = 0; i < 10; i++)
+    {
+        printf("%s %d \n", msg, i);
+        sleep(2);
+    }
+
+    return NULL;
+}
+```
+
+To compile the above code, you have to use the following command:
+
+```shell
+chmod u+x filename.c
+gcc filename.c -o pthread_example -pthread
+./pthread_example
+```
+
+The output for the above code will be:
+
+```shell
+Threading2 % ls
+message.c
+
+Threading2 % chmod u+x message.c 
+Threading2 % gcc message.c -o pthread_example -pthread
+
+Threading2 % ls
+message.c	pthread_example
+
+Threading2 % ./pthread_example 
+Main function after pthread_create 
+first thread 0 
+second thread 0 
+first thread 1 
+second thread 1 
+first thread 2 
+second thread 2 
+first thread 3 
+second thread 3 
+second thread 4 
+first thread 4 
+second thread 5 
+first thread 5 
+second thread 6 
+first thread 6 
+second thread 7 
+first thread 7 
+second thread 8 
+first thread 8 
+second thread 9 
+first thread 9 
+first thread ret1 = 0 
+second thread ret2 = 0 
+```
+
+There will be 2 seconds of gap between each output and it is due to `sleep(2)` function call inside the `myfunc` function. The purpose of `sleep(2)` is to simulate some processing or work being done within each iteration of the loop. By introducing this delay, you can observe the interleaved execution of the two threads more easily.
+
+## Types of Threads
+
+Threads are implemented in the following two ways −
+
+- User Level Threads
+  - User managed threads.
+  - These are the threads that application programmers use in their programs.
+- Kernel Level Threads
+  - Operating System managed threads acting on kernel.
+  - Kernel threads are supported within the kernel of the OS itself. All modern OSs support kernel level threads, allowing the kernel to perform multiple simultaneous tasks or to service multiple kernel system calls simultaneously.
+
+|  USER LEVEL THREAD |  KERNEL LEVEL THREAD   |
+|-----------|-----------|
+| User thread are implemented by users.  | kernel threads are implemented by OS.  |
+| OS doesn't recognized user level threads.  | Kernel threads are recognized by OS.  |
+| Implementation of User threads is easy.  | Implementation of Kernel thread is complicated.|
+| Context switch time is less.  | Context switch time is more.  |
+| Context switch requires no hardware support.| Hardware support is needed.  |
+| If one user level thread perform blocking operation then the entire process will be blocked.  | If one kernel thread perform blocking operation then another thread can continue execution.  |
+|Example: Java thread, POSIX threads.|Example: Window Solaris.|
+
+## Multithreading Models
+
+### The user threads must be mapped to kernel threads, by one of the following strategies:
+
+- Many to One Model
+- One to One Model
+- Many To Many Model
+
+## Many to One Model
```

Comment:
```suggestion
### Many to One Model
```

I think you meant to make these h3s nested under the h2 above

Reply:
Yes correct